File: frontend-web/src/services/notesService.js
/**
 * Service pour gérer les notes collaboratives
 */
import apiClient from './api';

export const notesService = {
  /**
   * Créer une nouvelle note
   */
  async createNote(title, folderId = null, content = '') {
    const response = await apiClient.post('/notes', {
      title,
      folder_id: folderId,
      content,
    });
    return response.data;
  },

  /**
   * Lister les notes
   */
  async listNotes(folderId = null, sharedWithMe = false) {
    const response = await apiClient.get('/notes', {
      params: {
        folder_id: folderId,
        shared_with_me: sharedWithMe,
      },
    });
    return response.data;
  },

  /**
   * Obtenir une note spécifique
   */
  async getNote(noteId) {
    const response = await apiClient.get(`/notes/${noteId}`);
    return response.data;
  },

  /**
   * Mettre à jour une note
   */
  async updateNote(noteId, updates) {
    const response = await apiClient.patch(`/notes/${noteId}`, updates);
    return response.data;
  },

  /**
   * Supprimer une note
   */
  async deleteNote(noteId) {
    const response = await apiClient.delete(`/notes/${noteId}`);
    return response.data;
  },

  /**
   * Restaurer une note
   */
  async restoreNote(noteId) {
    const response = await apiClient.post(`/notes/${noteId}/restore`);
    return response.data;
  },

  /**
   * Supprimer définitivement une note
   */
  async permanentDeleteNote(noteId) {
    const response = await apiClient.delete(`/notes/${noteId}/permanent`);
    return response.data;
  },

  /**
   * Partager une note avec un utilisateur
   */
  async shareNote(noteId, userId, permission = 'read') {
    const response = await apiClient.post(`/notes/${noteId}/share`, {
      user_id: userId,
      permission,
    });
    return response.data;
  },

  /**
   * Retirer le partage d'une note
   */
  async unshareNote(noteId, userId) {
    const response = await apiClient.post(`/notes/${noteId}/unshare`, {
      user_id: userId,
    });
    return response.data;
  },

  /**
   * Créer un lien public
   */
  async createPublicLink(noteId, options = {}) {
    const response = await apiClient.post(`/notes/${noteId}/public-link`, options);
    return response.data;
  },

  /**
   * Obtenir une note publique
   */
  async getPublicNote(token) {
    const response = await apiClient.get(`/notes/public/${token}`);
    return response.data;
  },
};

File: mobile-app/lib/services/notes_service.dart
import '../models/note.dart';
import 'api_service.dart';

class NotesService {
  final ApiService _apiService = ApiService();

  /// Créer une nouvelle note
  Future<Note> createNote(String title, {String? folderId, String? content}) async {
    try {
      final response = await _apiService.post(
        '/notes',
        data: {
          'title': title,
          'content': content ?? '',
          'folder_id': folderId,
        },
      );
      return Note.fromJson(response.data['note']);
    } catch (e) {
      throw Exception('Erreur lors de la création de la note: $e');
    }
  }

  /// Lister les notes de l'utilisateur
  Future<List<Note>> listNotes({String? folderId, bool sharedWithMe = false}) async {
    try {
      final response = await _apiService.get(
        '/notes',
        queryParameters: {
          if (folderId != null) 'folder_id': folderId,
          'shared_with_me': sharedWithMe.toString(),
        },
      );
      final notes = (response.data['notes'] as List)
          .map((e) => Note.fromJson(e))
          .toList();
      return notes;
    } catch (e) {
      throw Exception('Erreur lors du chargement des notes: $e');
    }
  }

  /// Obtenir une note spécifique
  Future<Note> getNote(String noteId) async {
    try {
      final response = await _apiService.get('/notes/$noteId');
      return Note.fromJson(response.data['note']);
    } catch (e) {
      throw Exception('Erreur lors du chargement de la note: $e');
    }
  }

  /// Mettre à jour une note
  Future<Note> updateNote(String noteId, {
    String? title,
    String? content,
    int? version,
  }) async {
    try {
      final data = <String, dynamic>{};
      if (title != null) data['title'] = title;
      if (content != null) data['content'] = content;
      if (version != null) data['version'] = version;

      final response = await _apiService.put(
        '/notes/$noteId',
        data: data,
      );
      return Note.fromJson(response.data['note']);
    } catch (e) {
      throw Exception('Erreur lors de la mise à jour de la note: $e');
    }
  }

  /// Supprimer une note
  Future<void> deleteNote(String noteId) async {
    try {
      await _apiService.delete('/notes/$noteId');
    } catch (e) {
      throw Exception('Erreur lors de la suppression de la note: $e');
    }
  }

  /// Restaurer une note
  Future<Note> restoreNote(String noteId) async {
    try {
      final response = await _apiService.post('/notes/$noteId/restore');
      return Note.fromJson(response.data['note']);
    } catch (e) {
      throw Exception('Erreur lors de la restauration de la note: $e');
    }
  }
}

